src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,116 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] storage;
+    private int size;
+
+    public ArrayList() {
+        storage = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= storage.length) {
+            grow();
+        }
+        storage[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (size >= storage.length) {
+            grow();
+        }
+        moveAdd(index);
+        storage[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public T remove(int index) {
+        checkIndex(index);
+        Object[] replaceStorage = new Object[storage.length];
+        System.arraycopy(storage, 0, replaceStorage, 0, index);
+        System.arraycopy(storage, index + 1, replaceStorage, index, (storage.length - index - 1));
+        T oldValue = storage[index];
+        updateStorage(replaceStorage);
+        size--;
+        return oldValue;
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public T remove(T element) {
+        for (int i = 0; i < size; i++) {
+            if (storage[i] == element || (storage[i] != null && storage[i].equals(element))) {
+                T oldValue = storage[i];
+                remove(i);
+                return oldValue;
+            }
+        }
+        throw new NoSuchElementException("Element not found");
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public T get(int index) {
+        checkIndex(index);
+        return storage[index];
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public void set(T value, int index) {
+        checkIndex(index);
+        storage[index] = value;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index must be: "
+                    + "index < size && index >= 0");
+        }
+    }
+
+    private void grow() {
+        int newLength = storage.length + (storage.length >> 1);
```

Comment:
what about adding RESIZE_FACTOR as a constant to make it more clear how we change the size of the array?